iothub_client/tests/common_e2e/iothubclient_common_e2e.c

```diff
@@ -681,56 +843,86 @@ D2C_MESSAGE_HANDLE client_create_and_send_d2c(TEST_MESSAGE_CREATION_MECHANISM te
         }
         ASSERT_IS_NOT_NULL(msgHandle, "Could not create the D2C message to be sent");
 
-        for (size_t i = 0; i < MSG_PROP_COUNT; i++)
+        for (size_t j = 0; j < MSG_PROP_COUNT; j++)
         {
             if (g_e2e_test_options.use_special_chars)
             {
-                if (IoTHubMessage_SetProperty(msgHandle, MSG_PROP_KEYS_SPECIAL[i], MSG_PROP_VALS_SPECIAL[i]) != IOTHUB_MESSAGE_OK)
+                if (IoTHubMessage_SetProperty(msgHandle, MSG_PROP_KEYS_SPECIAL[j], MSG_PROP_VALS_SPECIAL[j]) != IOTHUB_MESSAGE_OK)
                 {
-                    LogError("ERROR: Map_AddOrUpdate failed for property %zu!", i);
+                    LogError("ERROR: Map_AddOrUpdate failed for property %zu!", j);
                 }
             }
             else
             {
-                if (IoTHubMessage_SetProperty(msgHandle, MSG_PROP_KEYS[i], MSG_PROP_VALS[i]) != IOTHUB_MESSAGE_OK)
+                if (IoTHubMessage_SetProperty(msgHandle, MSG_PROP_KEYS[j], MSG_PROP_VALS[j]) != IOTHUB_MESSAGE_OK)
                 {
-                    LogError("ERROR: Map_AddOrUpdate failed for property %zu!", i);
+                    LogError("ERROR: Map_AddOrUpdate failed for property %zu!", j);
                 }
             }
         }
 
         sendData->msgHandle = msgHandle;
+        d2cMessages->items[i] = sendData;
+        d2cMessages->count++;
 
         // act
         sendeventasync_on_device_or_module(msgHandle, sendData);
     }
+}
 
-    return (D2C_MESSAGE_HANDLE)sendData;
+D2C_MESSAGE_HANDLE client_create_and_send_d2c(TEST_MESSAGE_CREATION_MECHANISM test_message_creation)
+{
+    SEND_DATA_BUNDLE d2cMessages;
+    client_create_and_send_d2c_messages(test_message_creation, 1, DEFAULT_MESSAGE_SIZE, &d2cMessages);
+    return d2cMessages.items[0];
 }
 
-bool client_wait_for_d2c_confirmation(D2C_MESSAGE_HANDLE d2cMessage, IOTHUB_CLIENT_CONFIRMATION_RESULT expectedClientResult)
+bool client_wait_for_d2c_confirmations(SEND_DATA_BUNDLE* d2cMessages, IOTHUB_CLIENT_CONFIRMATION_RESULT expectedClientResult)
 {
     time_t beginOperation, nowTime;
-    bool ret;
+    bool allMessagesSent = false;
```

Comment:
Nit - I think this is more like `allMessagesAcked` or something?  Since we could've sent the message technically but the underlying of this (on MQTT specifically) is a PUBACK.  

Though I don't know if that terminology makes sense in AMQP land?  Maybe allMessagesConfirmed to keep parallel with the function name here?